homeassistant/components/fan/comfoconnect.py

```diff
@@ -0,0 +1,278 @@
+            SENSOR_HUMIDITY_OUTDOOR
+        )
+
+        data = {}
+
+        for key, value in self._data.items():
+            if key == SENSOR_FAN_SPEED_MODE:
+                if value == 0:
+                    data[ATTR_SPEED] = SPEED_OFF
+                elif value == 1:
+                    data[ATTR_SPEED] = SPEED_LOW
+                elif value == 2:
+                    data[ATTR_SPEED] = SPEED_MEDIUM
+                elif value == 3:
+                    data[ATTR_SPEED] = SPEED_HIGH
+
+            elif key == SENSOR_HUMIDITY_EXTRACT:
+                data[ATTR_CURRENT_HUMIDITY] = value
+
+            elif key == SENSOR_HUMIDITY_OUTDOOR:
+                data[ATTR_OUTSIDE_HUMIDITY] = value
+
+            elif key == SENSOR_TEMPERATURE_EXTRACT:
+                data[ATTR_CURRENT_TEMPERATURE] = value / 10.0
+
+            elif key == SENSOR_TEMPERATURE_OUTDOOR:
+                data[ATTR_OUTSIDE_TEMPERATURE] = value / 10.0
+
+            elif key == SENSOR_FAN_SUPPLY_FLOW:
+                data[ATTR_AIR_FLOW_SUPPLY] = value
+
+            elif key == SENSOR_FAN_EXHAUST_FLOW:
+                data[ATTR_AIR_FLOW_EXTRACT] = value
+
+        return data
+
+    def update(self):
+        """Open connection to the Bridge."""
+        from pycomfoconnect.error import (
+            PyComfoConnectOtherSession, PyComfoConnectNotAllowed)
+
+        _LOGGER.debug('Updating sensor data...')
+        try:
+            self._comfoconnect.connect(
+                self._token, self._friendly_name, self._pin)
+            _LOGGER.debug('Connected to bridge.')
+
+            # Clean sensor data
+            self._fresh_data = {}
+
+            # Subscribe to sensor values.
+            for sensor in self._subscribed_sensors:
+                self._comfoconnect.request(sensor)
+            _LOGGER.debug('Subscribed to sensors.')
+
+            # Wait maximum of 5 seconds for all the sensor values.
+            self.wait_for_sensor_values(5)
+            _LOGGER.debug('Sensor data received.')
+
+            # Update data with new values
+            self._data.update(self._fresh_data)
+
+        except PyComfoConnectOtherSession as ex:
+            _LOGGER.error('Another session with "%s" is active.',
+                          ex.devicename)
+
+        except PyComfoConnectNotAllowed:
+            _LOGGER.error('Could not register. Invalid PIN!')
+
+        finally:
+            self._comfoconnect.disconnect()
+            _LOGGER.debug('Disconnected from bridge.')
+
+    def wait_for_sensor_values(self, max_seconds):
+        """Wait for all the sensor values to have arrived."""
+        deadline = time.time() + max_seconds
+        for sensor in self._subscribed_sensors:
+            if sensor in self._fresh_data:
+                continue
+
+            if time.time() > deadline:
+                _LOGGER.error('Timeout while waiting for sensor data.')
+                return
+
+            time.sleep(1)
```

Comment:
Home Assistant only has 10 threads to do work for synchronous components/platforms. Blocking a thread with a sleep is not allowed. Instead, use `track_point_in_time` to schedule a function to be called again